### PR TITLE
[AMDGPU] Shrink to SOPK with 32-bit signed literals

### PR DESCRIPTION
**Pull request by rampitec**
A literal like 0xffff8000 is valid to be used as KIMM in a SOPK instruction, but at the moment our checks expect it to be fully sign extended to a 64-bit signed integer. This is not required since all cases which are being shrunk only accept 32-bit operands.

We need to sign extend the operand to 64-bit though so it passes the verifier and properly printed.

### CODE REVIEW

File: llvm/lib/Target/AMDGPU/SIShrinkInstructions.cpp [hunk 1]
```diff
@@ -159,7 +159,7 @@ bool SIShrinkInstructions::shouldShrinkTrue16(MachineInstr &MI) const {
 }
 
 bool SIShrinkInstructions::isKImmOperand(const MachineOperand &Src) const {
-  return isInt<16>(Src.getImm()) &&
+  return isInt<16>(SignExtend64(Src.getImm(), 32)) &&
```
Comment (jayfoad) [inline llvm/lib/Target/AMDGPU/SIShrinkInstructions.cpp:162]: Could write this as `isInt<16>((int32_t)Src.getImm())` but it's a matter of taste I guess.

**Review by jayfoad — APPROVED** [review]
I'm a bit surprised it doesn't affect any codegen tests.

Anyway makes sense to me since the src operand of S_MOVK_I32 is signed but the src operand of S_MOV_B32 is not (B32 is neither signed nor unsigned, but I guess we generally treat it like unsigned).

**Comment by rampitec** [comment]
> I'm a bit surprised it doesn't affect any codegen tests.
> 
> Anyway makes sense to me since the src operand of S_MOVK_I32 is signed but the src operand of S_MOV_B32 is not (B32 is neither signed nor unsigned, but I guess we generally treat it like unsigned).

I was surprised too, but this is all because we do not select 64-bit moves. All of that is a step to select these.

File: llvm/lib/Target/AMDGPU/SIShrinkInstructions.cpp (hunk 1 above)
Comment (rampitec) [reply llvm/lib/Target/AMDGPU/SIShrinkInstructions.cpp:162]: > Could write this as `isInt<16>((int32_t)Src.getImm())` but it's a matter of taste I guess.

Same thing in principle.

File: llvm/test/CodeGen/AMDGPU/shrink-i32-kimm.mir [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+---
+name:            shrink_kimm32_mul_i32
+tracksRegLiveness: true
+body:             |
+  bb.0:
+
+    ; GCN-LABEL: name: shrink_kimm32_mul_i32
+    ; GCN: $sgpr0 = S_MULK_I32 undef $sgpr0, -2048, implicit-def $scc
+    $sgpr0 = S_MUL_I32 undef $sgpr0, 4294965248, implicit-def $scc
```
Comment (DadSchoorse) [inline llvm/test/CodeGen/AMDGPU/shrink-i32-kimm.mir:56]: Why do these have an implicit scc def? s_mul doesn't write scc.

Comment (jayfoad) [reply llvm/test/CodeGen/AMDGPU/shrink-i32-kimm.mir:56]: Good point, but this is just a harmless oversight in the handwritten MIR input. LLVM codegen knows that S_MUL_I32 doesn't write $scc.

Comment (rampitec) [reply llvm/test/CodeGen/AMDGPU/shrink-i32-kimm.mir:56]: Yeah, right. Result of the copy-paste in the handwritten MIR.

Comment (rampitec) [reply llvm/test/CodeGen/AMDGPU/shrink-i32-kimm.mir:56]: Fixed: https://github.com/llvm/llvm-project/pull/70489

**Pull request merged**